### PR TITLE
Support outputing optimized program after each update

### PR DESCRIPTION
**Pull request by RabbitWhite1**
With this, Flay can output optimized programs into a specified directory, using file names in the format of "optimized.{updateIdx}.p4".

### CODE REVIEW

**Comment by RabbitWhite1** [comment]
TODO: 
- [ ] use update file name to name output program.

File: core/flay_service.cpp [hunk 1]
```diff
@@ -119,8 +119,11 @@ std::pair<int, bool> FlayServiceBase::elimControlPlaneDeadCode(
                                : std::pair{EXIT_FAILURE, hasChanged};
 }
 
-void FlayServiceWrapper::outputOptimizedProgram(std::filesystem::path optimizedOutputFile) {
-    flayService.outputOptimizedProgram(optimizedOutputFile);
+void FlayServiceWrapper::outputOptimizedProgram(std::filesystem::path optimizedOutputFileName) {
+    auto absoluteFilePath =
+        FlayOptions::get().getOptimizedOutputDir().value() / optimizedOutputFileName;
+    flayService.outputOptimizedProgram(absoluteFilePath);
+    printInfo("Outputed optimized program to %1%", absoluteFilePath);
```
Comment (fruffy) [inline core/flay_service.cpp:126]:
```suggestion
    printInfo("Wrote optimized program to %1%", absoluteFilePath);
```

File: core/flay_service.cpp [hunk 2]
```diff
@@ -191,6 +195,10 @@ int FlayServiceWrapper::run() {
             recordProgramChange(flayService);
             semanticsChangeCounter++;
         }
+        if (FlayOptions::get().getOptimizedOutputDir() != std::nullopt) {
+            std::filesystem::path fileName = "optimized." + std::to_string(updateIdx) + ".p4";
```
Comment (fruffy) [inline core/flay_service.cpp:199]: You can use https://en.cppreference.com/w/cpp/filesystem/path/replace_extension here

**Review by fruffy — APPROVED** [review]
The failure happens because of some upstream changes.

**Review by fruffy — APPROVED** [review]

**Pull request merged**